lib/Doctrine/DBAL/Driver/PDOStatementImplementations.php

```diff
@@ -0,0 +1,140 @@
+<?php
+
+namespace Doctrine\DBAL\Driver;
+
+use Doctrine\DBAL\Driver\PDO\Exception;
+use PDOException;
+
+use function array_filter;
+use function assert;
+use function func_get_args;
+use function is_array;
+
+use const PHP_VERSION_ID;
+
+if (PHP_VERSION_ID >= 80000) {
+    /**
+     * @internal
+     */
+    trait PDOStatementImplementations
+    {
+        /**
+         * {@inheritdoc}
+         *
+         * @deprecated Use one of the fetch- or iterate-related methods.
+         */
+        public function setFetchMode($mode, ...$args)
+        {
+            $mode = $this->convertFetchMode($mode);
+
+            $filteredArgs = array_filter($args, static function (mixed $value): bool {
```

Comment:
Does the `mixed` type have to be used here? As far as I understand, it still gets parsed on PHP 7 but is interpreted as a class named `mixed` which is not what we want even if this code is not executed.

Reply:
Does this implementation have to be different from the one for PHP 7? If three arguments are passed, and the first two are NULL while the third is not, the value of the third argument will be passed as the first one.

Reply:
> Does the `mixed` type have to be used here?
No. It's pure eye candy because we're in a code branch that will only be executed in PHP 8. I can remove it if you don't like it. 😃 

> Does this implementation have to be different from the one for PHP 7?

We have different parameters, so the implementation is different. If the current logic is too confusing, I'm open to suggestions.

> If three arguments are passed, and the first two are NULL while the third is not, the value of the third argument will be passed as the first one.

No, what I'm checking here is if there are any additional non-null parameters passed. If that is the case, I'm passing the additional arguments as they are, otherwise I'm passing no additional arguments at all.

> UPD: I see that the original `$args` gets eventually passed down the line, so the intent is is even less clear.

The method signature might have changed, but PDO is still picky about passing unwanted `null` arguments.

Reply:
> We have different parameters, so the implementation is different.

The only change in PHP 8 I'm aware of is the signature change of `PDOStatement::query()` (https://github.com/doctrine/dbal/pull/), there must have been other signature changes (e.g. https://github.com/php/php-src/commit/4a5e17f9afedb6bd750ae8a89a0cf705f738c7e4). Please document all that changes that this patch attempts to adress.

Reply:
The methods that changed their signature in an incompatible way are
* `PDO::query`
* `PDOStatement::setFetchMode()`
* `PDOStatement::fetchAll()`

Apart from that, a lot of functions and methods provided by php have been upgraded with scalar type declarations. This change also affects PDO and is basically why I had to fix that one call to `bindParam()`.

Reply:
I have changed `array_filter` call, hoping that the new implementation makes the intent of that code block clearer.

Reply:
Please do not resolve a conversation started by someone else unless you're resolving it on their behalf.

Reply:
> The methods that changed their signature in an incompatible way [...] Apart from that, a lot of functions and methods provided by php have been upgraded with scalar type declarations.

It still doesn't answer why we need different implementations in order to satisfy just the difference in method signatures. I mean specifically handling `$args` in `setFetchMode()` but it may be more.

Reply:
The main difference between the two implementations is that the php 7 version of the signature allows up to two additional arguments while the php 8 signature has a variadic part. Both have in common that PDO will throw an error if additional arguments are passed although the selected fetch mode does not accept any.

The way that DBAL currently solves that problem is to only pass on the additional arguments is they aren't null. My attempt to preserve that logic was to search the variadic part for non-null values. If it doesn't contain any, I throw the variadic part away. If you know a better way, I'm open to suggestions.

Reply:
This is the approach I had in mind (the pseudo-code):
```php
if (PHP_VERSION_ID >= 80000) {
    trait PDOStatementInterface
    {
        use PDOStatementImplementation;

        public function fetchAll($mode = null, ...$args)
        {
            return $this->doFetchAll($mode, ...$args);
        }
    }
} else {
    trait PDOStatementInterface
    {
        use PDOStatementImplementation;

        public function fetchAll($fetchMode = null, $fetchArgument = null, $ctorArgs = null)
        {
            $args = func_get_args();
            array_shift($args);

            $this->doFetchAll($fetchMode, ...$args);
        }
    }
}

trait PDOStatementImplementation
{
    private function doFetchAll(...$args)
    {
        // the actual implementation
    }
}
```

Reply:
Yes, we can do this, I suppose. But I wouldn't create a separate trait for `doFetchAll()` but declare that method inside the actual `PDOStatement` class. I'll give it a try.

Reply:
I've pushed a commit. Please have a look.